analyzers/ImpactParameter.py

```diff
@@ -1,48 +1,233 @@
-                #                partic = const[i][1][l]
-                #                h = partic.path
-                #                Helix.compute_IP(h, assumed_vertex)
-                #                setattr(partic, self.IP, h.IP)
-                #                setattr(partic, self.tIP, h.closest.t)
-                #                setattr(partic, self.xIP, h.closest.x)
-                #                setattr(partic, self.yIP, h.closest.y)
-                #                setattr(partic, self.zIP, h.closest.z)
-                #                IPsn *= h.IP
-                #                N_charge += 1
-                #        else :
-                #            pass
-                #    else :
-                #        pass
-                #if N_charge <> 0 :
-                #    IPs = IPSn**(1.0/N_charge)
-                #else :
-                #    pass
+            IP_b_LL = 0
+            IPs_b_LL = 0
+            ptcs_pt = []
+            ptcs_IP = []
+            ptcs_IPxy = []
+            ptcs_IPz = []
+            ptcs_IP_signif = []
+            ptcs_IP_signif_sorted = []
+            ptcs_x_at_time_0 = []
+            ptcs_y_at_time_0 = []
+            ptcs_z_at_time_0 = []
+            ptcs_dr = []
             for id, ptcs in jet.constituents.iteritems():
-                if id in [22,130]:
+                if abs(id) in [22,130, 11]:
                     continue
                 for ptc in ptcs :
                     if ptc._charge == 0 :
                         continue
-                    ptc.path.compute_IP(assumed_vertex)
-                    #jet.IP = ptc.path.IP
-            #setattr(event, self.cfg_ana.output, IPs)
+                    ptc.path.compute_IP(assumed_vertex,jet)
+                    
+                    ptc_IP_signif = 0
+                    if hasattr(ptc.path, 'points') == True and 'beampipe_in' in ptc.path.points:
+                        ptc.path.compute_theta_0()
+                        ptc.path.compute_IP_signif(ptc.path.IP,
+                                                   ptc.path.theta_0,
+                                                   ptc.path.points['beampipe_in'])
+                    else :
+                        ptc.path.compute_IP_signif(ptc.path.IP, None, None)
+                    
+                    if ptc.path.p4.Perp() > 1 and (ptc.path.IPx**2 + ptc.path.IPy**2)**0.5 < 0.002 and ptc.path.IPz**2 < 0.17**2 :
+                        
+                        dr = deltaR(jet.eta(), jet.phi(), ptc.eta(), ptc.phi())
+                        
+                        ptcs_pt.append(ptc.path.p4.Perp())
+                        ptcs_IP.append(ptc.path.IP)
+                        ptcs_IPxy.append((ptc.path.IPx**2 + ptc.path.IPy**2)**0.5)
+                        ptcs_IPz.append(ptc.path.IPz)
+                        ptcs_IP_signif.append((ptc.path.IP_signif))
+                        ptcs_IP_signif_sorted.append((ptc.path.IP_signif))
```

Comment:
I used a second list to be able to get the other informations about the ptc, but your method seems to be much more efficient.